ggml/src/ggml-metal/ggml-metal.m

```diff
@@ -5892,37 +5899,172 @@ static void ggml_backend_metal_buffer_free_buffer(ggml_backend_buffer_t buffer)
+
+                [encoder copyFromBuffer:buf_src
+                           sourceOffset:0
+                               toBuffer:buf_dst
+                      destinationOffset:buf_dst_offset
+                                   size:size];
+
+                [encoder endEncoding];
+            }
+
+            [cmd_buf addCompletedHandler:^(id<MTLCommandBuffer> cb) {
+                // TODO: can check for errors here
+                GGML_UNUSED(cb);
+
+                dispatch_semaphore_signal(completion_semaphore);
+            }];
+
+            [cmd_buf commit];
+
+            dispatch_semaphore_wait(completion_semaphore, DISPATCH_TIME_FOREVER);
+            //[cmd_buf waitUntilCompleted];
+        }
+    }
 }
 
 static void ggml_backend_metal_buffer_get_tensor(ggml_backend_buffer_t buffer, const struct ggml_tensor * tensor, void * data, size_t offset, size_t size) {
-    memcpy(data, (const char *)tensor->data + offset, size);
+    struct ggml_backend_metal_buffer_context * ctx = (struct ggml_backend_metal_buffer_context *)buffer->context;
+
+    if (ctx->is_shared) {
+        memcpy(data, (const char *)tensor->data + offset, size);
+    } else {
+        @autoreleasepool {
+            // src
+            size_t buf_src_offset = 0;
+            id<MTLBuffer> buf_src = ggml_metal_get_buffer(tensor, &buf_src_offset);
+
+            buf_src_offset += offset;
+
+            // dst
+            id<MTLBuffer> buf_dst = [ctx->device newBufferWithBytesNoCopy:data
+                                                                   length:size
+                                                                  options:MTLResourceStorageModeShared
+                                                              deallocator:nil];
+
+            id<MTLCommandQueue>  queue   = ctx->queue;
+            id<MTLCommandBuffer> cmd_buf = [queue commandBufferWithUnretainedReferences];
+
+            {
+                id<MTLBlitCommandEncoder> encoder = [cmd_buf blitCommandEncoder];
+
+                [encoder copyFromBuffer:buf_src
+                           sourceOffset:buf_src_offset
+                               toBuffer:buf_dst
+                      destinationOffset:0
+                                   size:size];
 
-    GGML_UNUSED(buffer);
+                [encoder endEncoding];
+            }
+
+            [cmd_buf commit];
+            [cmd_buf waitUntilCompleted];
+        }
+    }
 }
 
 static bool ggml_backend_metal_buffer_cpy_tensor(ggml_backend_buffer_t buffer, const struct ggml_tensor * src, struct ggml_tensor * dst) {
     if (ggml_backend_buffer_is_host(src->buffer)) {
-        memcpy(dst->data, src->data, ggml_nbytes(src));
+        struct ggml_backend_metal_buffer_context * ctx = (struct ggml_backend_metal_buffer_context *)buffer->context;
+
+        if (ctx->is_shared) {
+            memcpy(dst->data, src->data, ggml_nbytes(src));
+        } else {
+            // TODO: implement
+            return false;
+        }
```

Comment:
It's not necessary to implement this, `ggml_backend_tensor_copy` already calls `ggml_backend_tensor_set` if the source tensor is in a host buffer.